### PR TITLE
Introduce SessionDecorator

### PR DESCRIPTION
**Pull request by kokosing**
Introduce SessionDecorator

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/dispatcher/DispatchManager.java [hunk 1]
```diff
@@ -169,6 +173,9 @@ public QueryId createQueryId()
 
             // decode session
             session = sessionSupplier.createSession(queryId, sessionContext);
+            if (sessionDecorator.isPresent()) {
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/dispatcher/DispatchManager.java:176]: there are way places where we create session e.g `io.prestosql.sql.analyzer.StatementAnalyzer#createViewSession` or 
`io.prestosql.dispatcher.DispatchManager#createQueryInternal`
What about these?

Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/dispatcher/DispatchManager.java:176]: > io.prestosql.sql.analyzer.StatementAnalyzer#createViewSession

Provides another decorated session (it wraps session with session).

> io.prestosql.dispatcher.DispatchManager#createQueryInternal

This is the place that you are commenting.

File: presto-main/src/main/java/io/prestosql/dispatcher/DispatchManager.java [hunk 2]
```diff
@@ -66,6 +67,7 @@
     private final TransactionManager transactionManager;
     private final AccessControl accessControl;
     private final SessionSupplier sessionSupplier;
+    private final Optional<SessionDecorator> sessionDecorator;
```
[the remark this answers is not in the document]
Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/dispatcher/DispatchManager.java:70]: Unfortunately, there is no option to override binding easily with guice. Alternatively, I could use `setBinder`.

File: presto-main/src/main/java/io/prestosql/server/SessionDecorator.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ * limitations under the License.
+ */
+
+package io.prestosql.server;
+
+import io.prestosql.Session;
+
+public interface SessionDecorator
+{
+    Session decorate(Session session, String query);
```
[the remark this answers is not in the document]
Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/server/SessionDecorator.java:21]: > It implies the orignal session is still kept around

I think this is good implication. Original has to be kept around as it contains very important information like `Identity`.

File: presto-main/src/main/java/io/prestosql/dispatcher/DispatchManager.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-main/src/main/java/io/prestosql/dispatcher/DispatchManager.java:70]: You could use something like: ```       
newOptionalBinder(binder, GlueColumnStatisticsProvider.class)
                .setDefault().to(DisabledGlueColumnStatisticsProvider.class).in(Scopes.SINGLETON);
```

File: presto-main/src/main/java/io/prestosql/server/SessionDecorator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/main/java/io/prestosql/server/SessionDecorator.java:21]: who's holding on to the original session?

**Pull request closed without merging**